src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -530,350 +581,792 @@ public static unsafe SafeCloseSocket CreateSocket(SocketInformation socketInform
 
-            return sent;
+            Interop.Error errno = Interop.Sys.GetLastError();
+            if (errno != Interop.Error.EAGAIN && errno != Interop.Error.EWOULDBLOCK)
+            {
+                errorCode = GetSocketErrorForErrorCode(errno);
+                return true;
+            }
+
+            errorCode = SocketError.Success;
+            return false;
         }
 
-        public static unsafe SocketError Receive(SafeCloseSocket handle, IList<ArraySegment<byte>> buffers, ref SocketFlags socketFlags, out int bytesTransferred)
+        public static unsafe bool TryStartConnect(int fileDescriptor, byte[] socketAddress, int socketAddressLen, out SocketError errorCode)
         {
-            var iovecs = new Interop.libc.iovec[buffers.Count];
-            var handles = new GCHandle[buffers.Count];
+            Debug.Assert(socketAddress != null);
+            Debug.Assert(socketAddressLen > 0);
 
-            try
+            int err;
+            fixed (byte* rawSocketAddress = socketAddress)
+            {
+                var sockAddr = (Interop.libc.sockaddr*)rawSocketAddress;
+                err = Interop.libc.connect(fileDescriptor, sockAddr, (uint)socketAddressLen);
+            }
+
+            if (err == 0)
+            {
+                errorCode = SocketError.Success;
+                return true;
+            }
+
+            Interop.Error errno = Interop.Sys.GetLastError();
+            if (errno != Interop.Error.EINPROGRESS)
+            {
+                errorCode = GetSocketErrorForErrorCode(errno);
+                return true;
+            }
+
+            errorCode = SocketError.Success;
+            return false;
+        }
+
+        public static unsafe bool TryCompleteConnect(int fileDescriptor, int socketAddressLen, out SocketError errorCode)
+        {
+            int socketErrno;
+			var optLen = (uint)sizeof(int);
+            int err = Interop.libc.getsockopt(fileDescriptor, Interop.libc.SOL_SOCKET, Interop.libc.SO_ERROR, &socketErrno, &optLen);
+
+            if (err != 0)
+            {
+                Debug.Assert(Interop.Sys.GetLastError() == Interop.Error.EBADF);
+                errorCode = SocketError.SocketError;
+                return true;
+            }
+			Debug.Assert(optLen == (uint)sizeof(int));
+
+            Interop.Error socketError = Interop.Sys.ConvertErrorPlatformToPal(socketErrno);
+            if (socketError == Interop.Error.SUCCESS)
+            {
+                errorCode = SocketError.Success;
+                return true;
+            }
+            else if (socketError == Interop.Error.EINPROGRESS)
+            {
+                errorCode = SocketError.Success;
+                return false;
+            }
+
+            errorCode = GetSocketErrorForErrorCode(socketError);
+
+// Disable CS0162: Unreachable code detected
+//
+// SuportsMultipleConnectAttempts is a constant; when false, the following lines will trigger CS0162.
+#pragma warning disable 162
```

Comment:
Instead of a code constant can't we have a `#define` and compile this section out?
It would remove the need for the `Debug.Fail()` within `PrimeForNextConnectAttempt` in OSX.